### ISSUE TITLE
Expose Excel parser and AST in R

### ISSUE THREAD

**Issue by nokome** [stencila/stencila#173]
There is some interest in having the Excel parser and generated AST exposed into R (e.g. https://github.com/stencila/stencila/issues/169#issuecomment-212317332).

This probably best exposed as a generic free function (i.e not attached to Sheets or any other component) which dispatches to whatever parsers may be implemented and returns an AST e.g.

``` R
ast <- parse_excel("SUM(A1:A10)")
```

What structure should be used for the AST? A nested list with `class` set on each node e.g. `Call`, `BinaryOp`. This could be implemented with a `ExcelToRAstGenerator` class which build up a list using Rcpp on calls to its `visit_xxx` functions.

**Comment by nokome** [comment]
Hi @jennybc and @richfitz : just pinging you here to let you know I have made a start on this. I have exposed the Excel formula parser into R with two functions

The `excel_ast` function parses Excel and converts the C++ AST into an R object. I have provided two forms - the default short form and a longer form which is more uniformly structured and might be better suited for analysis of the AST. 

```r
> str(excel_ast('SUM(A1:A10)'))
List of 2
 $ function : chr "SUM"
 $ arguments:List of 1
  ..$ :List of 2
  .. ..$ first: symbol A1
  .. ..$ last : symbol A10
  .. ..- attr(*, "class")= chr "Range"
 - attr(*, "class")= chr "Call"
> str(excel_ast('SUM(A1:A10)',mode='l'))
List of 3
 $ function : chr "SUM"
 $ arguments:List of 1
  ..$ :List of 3
  .. ..$ first:List of 2
  .. .. ..$ type : chr "Identifier"
  .. .. ..$ value: chr "A1"
  .. ..$ last :List of 2
  .. .. ..$ type : chr "Identifier"
  .. .. ..$ value: chr "A10"
  .. ..$ type : chr "Range"
 $ type     : chr "Call"
```

There is also a `excel_r` function:

```r
> eval(parse(text=excel_r('MODE(1,2,2,3)')))
[1] 2
```

More documentation and code here: https://github.com/stencila/stencila/blob/develop/r/stencila/syntax.R and here: https://github.com/stencila/stencila/blob/develop/r/stencila/syntax.cpp

The Excel parser itself is still incomplete. But in the meantime it would be great to get your thoughts on the above R interface. It's not something that will be used internally by Stencila so feel free to suggest changes to make it suit your needs.

Also, Jenny, I have got pretty close to getting a Mac OSX build #178 but really to get my hands on a local Mac (real or VM) to get the final bits sorted out.
